subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/MetaInfAwareClasspathResourceHasher.java

```diff
@@ -16,53 +16,82 @@
+import java.util.stream.Collectors;
+
+import static org.apache.commons.io.FileUtils.readFileToByteArray;
 
 public class MetaInfAwareClasspathResourceHasher implements ResourceHasher {
     private static final Logger LOGGER = LoggerFactory.getLogger(MetaInfAwareClasspathResourceHasher.class);
 
     private final ResourceHasher delegate;
-    private final ManifestFileZipEntryHasher manifestFileZipEntryHasher;
+    private final ResourceEntryFilter attributeResourceFilter;
 
-    public MetaInfAwareClasspathResourceHasher(ResourceHasher delegate, ManifestFileZipEntryHasher manifestFileZipEntryHasher) {
+    public MetaInfAwareClasspathResourceHasher(ResourceHasher delegate, ResourceEntryFilter attributeResourceFilter) {
         this.delegate = delegate;
-        this.manifestFileZipEntryHasher = manifestFileZipEntryHasher;
+        this.attributeResourceFilter = attributeResourceFilter;
     }
 
     @Override
     public void appendConfigurationToHasher(Hasher hasher) {
         delegate.appendConfigurationToHasher(hasher);
-        manifestFileZipEntryHasher.appendConfigurationToHasher(hasher);
+        attributeResourceFilter.appendConfigurationToHasher(hasher);
     }
 
     @Nullable
     @Override
     public HashCode hash(RegularFileSnapshotContext snapshotContext) {
-        return delegate.hash(snapshotContext);
+        String relativePath = String.join("/", snapshotContext.getRelativePathSegments().get());
+        if (isManifestFile(relativePath)) {
+            return tryHashWithFallback(snapshotContext);
+        } else {
+            return delegate.hash(snapshotContext);
+        }
     }
 
     @Nullable
     @Override
     public HashCode hash(ZipEntryContext zipEntryContext) throws IOException {
         ZipEntry zipEntry = zipEntryContext.getEntry();
         if (isManifestFile(zipEntry.getName())) {
-            return tryHashWithFallback(zipEntryContext, manifestFileZipEntryHasher);
+            return tryHashWithFallback(zipEntryContext);
         } else {
             return delegate.hash(zipEntryContext);
         }
     }
 
     @Nullable
-    private HashCode tryHashWithFallback(ZipEntryContext zipEntryContext, ZipEntryHasher hasher) throws IOException {
+    private HashCode tryHashWithFallback(RegularFileSnapshotContext snapshotContext) {
+        try {
+            return hashManifest(readFileToByteArray(new File(snapshotContext.getSnapshot().getAbsolutePath())));
+        } catch (IOException e) {
+            LOGGER.debug("Could not load fingerprint for " + snapshotContext.getSnapshot().getAbsolutePath() + ". Falling back to full entry fingerprinting", e);
+            return delegate.hash(snapshotContext);
+        }
+    }
+
+    @Nullable
+    private HashCode tryHashWithFallback(ZipEntryContext zipEntryContext) throws IOException {
         try {
-            return hasher.hash(zipEntryContext);
+            return hashManifest(ByteStreams.toByteArray(zipEntryContext.getEntry().getInputStream()));
```

Comment:
This looks like a change that could introduce some performance regression. We used to hash zip entries by processing the input stream. Now instead we read all of the content into a new `byte[]` (only to convert it back to an input stream). We should instead pass the input stream directly. We should pass a `FileInputStream` from the `RegularFileSnapshotContext` variant of this method, too.

Reply:
Good catch.  Looks like it's been this way since the manifest normalization was introduced.  Not sure why a byte stream was used here in the first place.  It's been switched to just pass the input stream directly instead.